Convert JIT/Methodical/Array/range Pri0 tests and hook up mobile testing

jkoritzinsky opened the pull request:
Convert a set of il-based Pri0 tests to use the new testing framework and update the runtime test tree's "send to Helix" infra to correctly support merged test payloads for mobile platforms.

trylek committed:
Experimentally convert JIT/Methodical/Array/range Pri0 tests

jkoritzinsky committed:
Get Android app publishing with the libraries test infra and the existing console-based runner functioning.

jkoritzinsky committed:
Generate the RunTests script for local test execution and write up MSBuild to support the same mechanisms as the libraries tests for local mobile testing.

jkoritzinsky committed:
Clean up extraneous RID setting

jkoritzinsky committed:
Update the individual tests for the range suite to have unique assembly and class names.

jkoritzinsky committed:
Implement xharness-based runner support for mobile platforms using local xharness build

jkoritzinsky committed:
Make SkipCategories a no-op instead of an exception throw as it's used on WASM startup.

jkoritzinsky committed:
Don't copy the Xunit-based runners when not using the standard xunit runner. Copy the rest ofthe referenced/used assemblies to the output directory. Fix assembly names to match the IL names.

jkoritzinsky committed:
Use published xharness version

jkoritzinsky committed:
Update merged runner payload testing for mobile/browser platforms.

jkoritzinsky committed:
Don't push up CORE_ROOT for mobile test legs. We don't use it there.

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into src-tests-merged-payload-mobile

jkoritzinsky commented:
/azp run runtime-manual

trylek commented on src/tests/Common/helixpublishwitharcade.proj:
```diff
@@ -429,61 +503,86 @@
+  <ItemGroup Condition="'$(TargetsMobile)' != 'true'">
     <MergedPayloads Update="@(MergedPayloads)">
+      <PayloadHelixCommandFile>$(MergedPayloadsRootDirectory)%(FileName)%(Extension)\HelixCommand.txt</PayloadHelixCommandFile>
       <MergedTestHelixCommand>$([System.IO.File]::ReadAllText('%(PayloadHelixCommandFile)'))</MergedTestHelixCommand>
```
Does this actually work for you? When I was working on the initial implementation of the merged test wrappers, it seemed to me that within the ItemGroup Update loop you only see the pre-existing metadata values i.o.w. to consume a previously calculated metadata I had to use two subsequent Update loops. It can easily be the case I was doing something wrong, I don't have any exact documentation of this behavior at my disposal, just please double-check this truly behaves the way you expect it to.

jkoritzinsky replied:
I'll check if this actually works and revert it if it doesn't. I was seeing success, but I'll keep an eye on the WASM leg to make sure.

jkoritzinsky replied:
I've validated that this works correctly.

trylek commented on src/tests/JIT/Methodical/Arrays/range/float64_range1.il:
```diff
@@ -20,19 +20,13 @@
-  
 }
 .assembly extern xunit.core {}
-.module 'test.exe'
```
Is this a remaining cleanup we should apply to all ilproj tests (putting .assembly in sync with the source file, removing the .module line)? I believe I should be able to tweak my IL rewriter to bulk process all runtime ilproj tests like this.

jkoritzinsky replied:
Yes, this cleanup is actually required for ProjectReferences to il projects to work as expected. Otherwise, the SDK's conflict resolution will throw away all but one referenced assembly.

trylek reviewed (approved):
Looks awesome to me, thanks Jeremy! I'd love to hear whether there are any general patterns we might be able to bulk transform to streamline subsequent refactorings, .assembly and .module seem obvious, my ILTransform also already supports quite a few tricks regarding class name / namespace deduplication or other details. Moving forward, I would also love to give another thought to the _d / _do / _r / _ro suffixes and some way to remove them by somehow encoding the desire to build & test the different codegen options in the metadata but I'm not yet fully there.

jkoritzinsky committed:
Move tests.props earlier so the SDK gets the RuntimeIdentifier property setting where it's expected.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Explicitly set BuildAsStandalone to false for the merged test runner assemblies.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Merge branch 'main' into src-tests-merged-payload-mobile

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Move the check for TestArchiveTestsDir to the target that actually uses it (since the target is conditional but the check wasnt)

jkoritzinsky committed:
Fix "no build" test handling for merged mobile test runs.

jkoritzinsky committed:
Merge branch 'src-tests-merged-payload-mobile' of github.com:jkoritzinsky/runtime into src-tests-merged-payload-mobile

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Exclude scripts in the AppBundle folder from the legacy runner test script discovery mechanism.

jkoritzinsky committed:
Fix assembly name calculation for range IL tests.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Merge branch 'main' into src-tests-merged-payload-mobile

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Try 2 to remove the AppBundle scripts from execution.

Also supress the warning about no source files for merged test assemblies, that's expected.

jkoritzinsky committed:
Merge branch 'src-tests-merged-payload-mobile' of github.com:jkoritzinsky/runtime into src-tests-merged-payload-mobile

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Fix typo

jkoritzinsky committed:
Another attempt to fix the incorrect script inclusion issue.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Use PayloadGroup metadata as the name for the merged apps test runs as we provide the correctly-named metadata property to point to the app.

jkoritzinsky commented:
/azp run runtime-manual

naricc commented on src/tests/Common/helixpublishwitharcade.proj:
```diff
@@ -197,9 +197,20 @@
     </ItemGroup>
 
+    <PropertyGroup Condition="'$(TargetsMobile)' == 'true' or '$(TargetsBrowser)' == 'true'">
+      <!-- When targeting mobile or browser runs, we publish self-contained builds and build mobile/wasm apps, so we want to exclude these files from the legacy payloads as well -->
```
I am kind of confused why we want to exclude these files? The ones in publish I understand, but why the AppBundle?

jkoritzinsky replied:
We run the merged payloads in a completely separate mechanism from how the legacy payloads run, so we don't want to accidentally include the merged payloads (including the apps that we're actually going to run on the device) in the legacy payloads. The AppBundle folders would only serve to bloat the zip we send over to Helix, which we don't want to do.

jkoritzinsky committed:
The legacy payloads use Core_Root to get the xunit console runner, so include it when we have any legacy payloads.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Fix batching so we can have multiple merged payloads

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Remove the condition on including coreroot. We discover the legacy payloads later and I don't want to easl with putting things in weird locations.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Fix pattern for finding apps for apple to check the correct path.

jkoritzinsky committed:
Make sure the line we write for running the wasm tests is an executable path from a shell.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky commented:
Marking this as ready for review as it seems the tests I've migrated over are correctly running in the mobile legs.

jkoritzinsky committed:
The app bundle output targetos is the xcode SDK, not the TargetOS we use in our build.

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into src-tests-merged-payload-mobile

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky committed:
Don't append the RID to the output path to avoid breaking issues.targets glob paths.

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky commented:
I've validated that all of the new merged runners are running correctly on mobile as well as desktop platforms.

This PR is ready for another round of review.

cc: @trylek @naricc @fanyang-mono

trylek commented on src/tests/Common/helixpublishwitharcade.proj:
```diff
@@ -429,61 +517,86 @@
     </HelixWorkItem>
 
-    <HelixWorkItem Include="@(MergedPayloads->Metadata('PayloadGroup'))" Condition=" '$(IsRunningOnMobileTargets)' == 'false' ">
+    <HelixWorkItem Include="@(MergedPayloads->Metadata('PayloadGroup'))" Condition=" '$(TargetsMobile)' != 'true' or '$(TargetsBrowser)' == 'true' ">
```
Nit: I am seeing at least four places that query this particular combination of conditions - ("browser or not mobile" and its negation "non-browser on mobile"), would it make sense to synthesize a new property for this purpose?

jkoritzinsky replied:
I can't think of a good name for a consolidated property, so I'd prefer to keep using our existing properties as-is, at least for now.

naricc replied:
Is what we are really testing here whether or not the target uses a custom host (i.e. something other than corerun?) We could call it that; "TargetUsesCustomHost"?

jkoritzinsky replied:
I've made this condition `TargetHasHelixXHarnessSdkSupport` as that's what best describes most of the conditions where this condition or its inverse is used.

trylek reviewed (approved):
Looks great to me, thank you! I admit I'm still not an expert on device testing so that side of your change certainly merits Nathan's and / or Fan's review.

jkoritzinsky committed:
Refactor Mobile/Browser conditions.

jkoritzinsky committed:
Merge branch 'main' into src-tests-merged-payload-mobile

jkoritzinsky commented:
/azp run runtime-manual

jkoritzinsky commented:
Failing legs aren't runtime test legs, so I'm going to merge this in.

fanyang-mono commented:
It seems that with the conversion, the msbuild task to create the mobile apps failed to be triggered. For example, this task https://github.com/dotnet/runtime/blob/main/src/tests/build.proj#L189

jkoritzinsky commented:
@fanyang-mono it looks like that target is correctly triggered for the "legacy" payloads still: https://dev.azure.com/dnceng/public/_build/results?buildId=1539343&view=logs&j=c1e461de-48bf-5e60-0a97-1d3931972820&t=ca8eb547-1fee-51dd-a4ab-6571ac77788b

The merged payloads don't use that target and instead use the same mechanism that the libraries tests use to build and run android test assemblies (iOS and WASM are similar).

fanyang-mono commented:
@jkoritzinsky so what would be the command to build and run these converted tests locally?

jkoritzinsky commented:
I believe you can run the `Test` MSBuild target on the merged wrapper project to run the test, same process as the libraries tests.
